src/vm_pmu.rs

```diff
@@ -368,4 +373,149 @@ impl VmPmuState {
             }
         }
     }
+
+    /// Helper function to call either SBI, or the internal implementation
+    /// for configure_matching_counters() depending on whether Sscdeleg is implemented
+    pub fn configure_matching_counters(
+        &self,
+        counter_index: u64,
+        counter_mask: u64,
+        config_flags: PmuCounterConfigFlags,
+        event_type: PmuEventType,
+        event_data: u64,
+    ) -> SbiResult<u64> {
+        if !CpuInfo::get().has_sscdeleg() {
+            sbi::api::pmu::configure_matching_counters(
+                counter_index,
+                counter_mask,
+                config_flags,
+                event_type,
+                event_data,
+            )
+        } else {
+            fn is_firmware_counter(event_type: PmuEventType) -> bool {
+                matches!(event_type, PmuEventType::Firmware(_))
+            }
+
+            fn is_fixed_counter(counter_index: usize) -> bool {
+                // CYCLE and INSRET are fixed counters
+                counter_index == 0 || counter_index == 2
+            }
+
+            fn is_timer(counter_index: u64) -> bool {
+                counter_index == 1
+            }
+
+            if is_firmware_counter(event_type) {
+                return Err(SbiError::InvalidParam);
+            }
+
+            let pmu_info = pmu::PmuInfo::get()?;
+            let counter_mask = pmu_info.filter_counter_mask(counter_index, counter_mask)?;
+            let index = if !config_flags.is_skip_match() {
+                let mut bitmask_iter = CounterMaskIter::new(counter_index, counter_mask);
+                bitmask_iter
+                    .find(|i| {
+                        matches!(self.counter_state[*i], NotConfigured) || is_fixed_counter(*i)
+                    })
+                    .map_or_else(|| Err(SbiError::InvalidParam), |i| Ok(i as u64))
+            } else {
+                if is_timer(counter_index) {
+                    return Err(SbiError::InvalidParam);
+                }
```

Comment:
Why only check this in the skip_match case? Shouldn't we be checking this always?

Reply:
Since the mask gets filtered upstream, we should never have a case where the iterator returns an index of 1. Note that it's OK to start with a counter_index of 1 as long as the counter_mask doesn't select the bit (so say something like counter_index == 1 && counter_mask = 0x10 is fine). Since we already check the same for skip_match above, maybe we should just remove the check?

    ```
/// Returns a filtered counter_mask if the PMU counter range can be configured.
    pub fn get_configurable_counter_range(
        &self,
        counter_index: u64,
        counter_mask: u64,
        config_flags: PmuCounterConfigFlags,
    ) -> SbiResult<u64> {
        use PmuCounterState::*;
        let pmu_info = pmu::PmuInfo::get()?;
        let counter_mask = pmu_info.filter_counter_mask(counter_index, counter_mask)?;
        if !config_flags.is_skip_match() {
            let mut bitmask_iter = CounterMaskIter::new(counter_index, counter_mask);
            bitmask_iter
                .find(|i| matches!(self.counter_state[*i], NotConfigured))
                .map_or_else(|| Err(SbiError::InvalidParam), |_| Ok(counter_mask))
        } else {
            // If skip_match is set, the counter must already be configured
            let state = self
                .counter_state
                .get(counter_index as usize)
                .ok_or(SbiError::InvalidParam)?;
            if matches!(state, Started(_) | Configured(_)) {
                Ok(counter_mask)
            } else {
                Err(SbiError::InvalidParam)
            }
        }
    }
```

Reply:
Yeah ok, if this is already accounted for then I'd suggest removing the check. It's confusing to have it one branch but not the other.